### ISSUE TITLE
Unable to retrieve results for e2e plugin

### ISSUE THREAD

**Issue by mauricewittek** [vmware-tanzu/sonobuoy#1508]
I'm running sonobuoy in an airgapped environment with the following parameters: 
`sonobuoy run --wait --mode quick --sonobuoy-image myrepo/sonobuoy/sonobuoy:main --kube-conformance-image myrepo/conformance:v1.19.15 --systemd-logs-image myrepo/heptio-images/sonobuoy-plugin-systemd-logs:latest --e2e-repo-config custom-repos.yaml`

This prints the following output:

```
time="2021-11-04T15:23:16+01:00" level=info msg="create request issued" name=sonobuoy namespace= resource=namespaces
time="2021-11-04T15:23:16+01:00" level=info msg="create request issued" name=sonobuoy-serviceaccount namespace=sonobuoy resource=serviceaccounts
time="2021-11-04T15:23:16+01:00" level=info msg="create request issued" name=sonobuoy-serviceaccount-sonobuoy namespace= resource=clusterrolebindings
time="2021-11-04T15:23:16+01:00" level=info msg="create request issued" name=sonobuoy-serviceaccount-sonobuoy namespace= resource=clusterroles
time="2021-11-04T15:23:16+01:00" level=info msg="create request issued" name=sonobuoy-config-cm namespace=sonobuoy resource=configmaps
time="2021-11-04T15:23:16+01:00" level=info msg="create request issued" name=sonobuoy-plugins-cm namespace=sonobuoy resource=configmaps
time="2021-11-04T15:23:16+01:00" level=info msg="create request issued" name=sonobuoy namespace=sonobuoy resource=pods
time="2021-11-04T15:23:16+01:00" level=info msg="create request issued" name=plugin-e2e-cm namespace=sonobuoy resource=configmaps
time="2021-11-04T15:23:16+01:00" level=info msg="create request issued" name=sonobuoy-aggregator namespace=sonobuoy resource=services
15:23:36          PLUGIN                                   NODE    STATUS   RESULT           PROGRESS
15:23:36             e2e                                 global   running            0/1 (0 failures)
15:23:36    systemd-logs   01c38d0d-e51a-43bd-81a7-1b06825c81e9   running
15:23:36    systemd-logs   2b3c43f4-da94-4af5-8c44-f1dc51cf0b72   running
15:23:36    systemd-logs   efe0f9ba-ce81-48ec-8013-02dd45074e85   running
15:23:36
15:23:36 Sonobuoy is still running. Runs can take 60 minutes or more depending on cluster and plugin configuration.
15:23:56             e2e                                 global   complete            1/1 (0 failures)
15:23:56    systemd-logs   01c38d0d-e51a-43bd-81a7-1b06825c81e9   complete
15:23:56    systemd-logs   2b3c43f4-da94-4af5-8c44-f1dc51cf0b72   complete
15:23:56    systemd-logs   efe0f9ba-ce81-48ec-8013-02dd45074e85   complete
15:23:56 Sonobuoy plugins have completed. Preparing results for download.

```

After the run is finished, `sonobuoy status` prints the following output

```
sonobuoy status
         PLUGIN     STATUS    RESULT   COUNT           PROGRESS
            e2e   complete   unknown       1   1/1 (0 failures)
   systemd-logs   complete    passed       3

Sonobuoy has completed. Use `sonobuoy retrieve` to get results.
```

I had a look at the logs with `kubectl logs sonobuoy -n sonobuoy`:

```
time="2021-11-04T14:39:34Z" level=info msg="Scanning plugins in ./plugins.d (pwd: /)"
time="2021-11-04T14:39:34Z" level=info msg="Scanning plugins in /etc/sonobuoy/plugins.d (pwd: /)"
time="2021-11-04T14:39:34Z" level=info msg="Directory (/etc/sonobuoy/plugins.d) does not exist"
time="2021-11-04T14:39:34Z" level=info msg="Scanning plugins in ~/sonobuoy/plugins.d (pwd: /)"
time="2021-11-04T14:39:34Z" level=info msg="Directory (~/sonobuoy/plugins.d) does not exist"
time="2021-11-04T14:39:35Z" level=info msg="Filtering namespaces based on the following regex:.*"
time="2021-11-04T14:39:35Z" level=info msg="Namespace default Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace deployment-4723 Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace ingress-nginx Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace kube-node-lease Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace kube-public Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace kube-system Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace nsx-system Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace pks-system Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace secrets-6991 Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace secrets-762 Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace sonobuoy Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Namespace statefulset-2466 Matched=true"
time="2021-11-04T14:39:35Z" level=info msg="Starting server Expected Results: [{global e2e} {01c38d0d-e51a-43bd-81a7-1b06825c81e9 systemd-logs} {2b3c43f4-da94-4af5-8c44-f1dc51cf0b72 systemd-logs} {efe0f9ba-ce81-48ec-8013-02dd45074e85 systemd-logs}]"
time="2021-11-04T14:39:35Z" level=info msg="Starting annotation update routine"
time="2021-11-04T14:39:35Z" level=info msg="Starting aggregation server" address=0.0.0.0 port=8080
time="2021-11-04T14:39:35Z" level=info msg="Running plugin" plugin=e2e
time="2021-11-04T14:39:35Z" level=info msg="Running plugin" plugin=systemd-logs
time="2021-11-04T14:39:38Z" level=info msg="received request" client_cert="[systemd-logs]" method=PUT node=2b3c43f4-da94-4af5-8c44-f1dc51cf0b72 plugin_name=systemd-logs url=/api/v1/results/by-node/2b3c43f4-da94-4af5-8c44-f1dc51cf0b72/systemd-logs
time="2021-11-04T14:39:38Z" level=info msg="received request" client_cert="[systemd-logs]" method=PUT node=01c38d0d-e51a-43bd-81a7-1b06825c81e9 plugin_name=systemd-logs url=/api/v1/results/by-node/01c38d0d-e51a-43bd-81a7-1b06825c81e9/systemd-logs
time="2021-11-04T14:39:39Z" level=info msg="received request" client_cert="[systemd-logs]" method=PUT node=efe0f9ba-ce81-48ec-8013-02dd45074e85 plugin_name=systemd-logs url=/api/v1/results/by-node/efe0f9ba-ce81-48ec-8013-02dd45074e85/systemd-logs
time="2021-11-04T14:39:42Z" level=info msg="received request" client_cert="[e2e]" method=POST plugin_name=e2e url=/api/v1/progress/global/e2e
time="2021-11-04T14:39:54Z" level=info msg="received request" client_cert="[e2e]" method=POST plugin_name=e2e url=/api/v1/progress/global/e2e
time="2021-11-04T14:39:54Z" level=info msg="received request" client_cert="[e2e]" method=POST plugin_name=e2e url=/api/v1/progress/global/e2e
time="2021-11-04T14:39:55Z" level=info msg="received request" client_cert="[e2e]" method=PUT plugin_name=e2e url=/api/v1/results/global/e2e
time="2021-11-04T14:39:55Z" level=error msg="Result processing error (500): error handling result e2e/global: couldn't decode result e2e/results/global: couldn't uncompress reader: gzip: invalid header"
time="2021-11-04T14:39:56Z" level=info msg="received request" client_cert="[e2e]" method=PUT plugin_name=e2e url=/api/v1/results/global/e2e
time="2021-11-04T14:39:56Z" level=error msg="Result processing error (500): error handling result e2e/global: couldn't decode result e2e/results/global: couldn't uncompress reader: gzip: invalid header"
time="2021-11-04T14:39:57Z" level=info msg="received request" client_cert="[e2e]" method=PUT plugin_name=e2e url=/api/v1/results/global/e2e
time="2021-11-04T14:39:57Z" level=error msg="Result processing error (500): error handling result e2e/global: couldn't decode result e2e/results/global: couldn't uncompress reader: gzip: invalid header"
time="2021-11-04T14:40:04Z" level=info msg="All plugins have completed, status has been updated"
time="2021-11-04T14:41:55Z" level=info msg="Shutting down aggregation server"
time="2021-11-04T14:41:55Z" level=info msg="Resources is not set explicitly implying query all resources, but skipping secrets for safety. Specify the value explicitly in Resources to gather this data."
time="2021-11-04T14:41:55Z" level=info msg="Collecting Node Configuration and Health..."
time="2021-11-04T14:41:55Z" level=info msg="Creating host results for 01c38d0d-e51a-43bd-81a7-1b06825c81e9 under /tmp/sonobuoy/dae67975-cdc1-488c-b11f-81ae10346043/hosts/01c38d0d-e51a-43bd-81a7-1b06825c81e9\n"
time="2021-11-04T14:41:55Z" level=info msg="Creating host results for 2b3c43f4-da94-4af5-8c44-f1dc51cf0b72 under /tmp/sonobuoy/dae67975-cdc1-488c-b11f-81ae10346043/hosts/2b3c43f4-da94-4af5-8c44-f1dc51cf0b72\n"
time="2021-11-04T14:41:55Z" level=info msg="Creating host results for efe0f9ba-ce81-48ec-8013-02dd45074e85 under /tmp/sonobuoy/dae67975-cdc1-488c-b11f-81ae10346043/hosts/efe0f9ba-ce81-48ec-8013-02dd45074e85\n"
time="2021-11-04T14:41:55Z" level=info msg="Running cluster queries"
W1104 14:41:55.832138       1 warnings.go:70] v1 ComponentStatus is deprecated in v1.19+
time="2021-11-04T14:41:55Z" level=info msg="Running ns query (default)"
W1104 14:41:56.008676       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:41:56Z" level=info msg="Running ns query (deployment-4723)"
W1104 14:41:56.326230       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:41:56Z" level=info msg="Running ns query (ingress-nginx)"
W1104 14:41:57.028112       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:41:57Z" level=info msg="Running ns query (kube-node-lease)"
W1104 14:41:58.028582       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:41:58Z" level=info msg="Running ns query (kube-public)"
W1104 14:41:59.027987       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:41:59Z" level=info msg="Running ns query (kube-system)"
W1104 14:42:00.197405       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:42:00Z" level=info msg="Running ns query (nsx-system)"
W1104 14:42:01.028426       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:42:01Z" level=info msg="Running ns query (pks-system)"
W1104 14:42:02.027309       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:42:02Z" level=info msg="Running ns query (secrets-6991)"
W1104 14:42:03.028056       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:42:03Z" level=info msg="Running ns query (secrets-762)"
W1104 14:42:04.028438       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:42:04Z" level=info msg="Running ns query (sonobuoy)"
W1104 14:42:05.028365       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:42:05Z" level=info msg="Running ns query (statefulset-2466)"
W1104 14:42:06.030214       1 warnings.go:70] extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
time="2021-11-04T14:42:06Z" level=info msg="Namespace default Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Namespace deployment-4723 Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Namespace ingress-nginx Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Namespace kube-node-lease Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Namespace kube-public Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Namespace kube-system Matched=true"
time="2021-11-04T14:42:06Z" level=info msg="Namespace nsx-system Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Namespace pks-system Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Namespace secrets-6991 Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Namespace secrets-762 Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Namespace sonobuoy Matched=true"
time="2021-11-04T14:42:06Z" level=info msg="Namespace statefulset-2466 Matched=false"
time="2021-11-04T14:42:06Z" level=info msg="Collecting Pod Logs by namespace (kube-system)"
time="2021-11-04T14:42:07Z" level=info msg="Collecting Pod Logs by namespace (sonobuoy)"
time="2021-11-04T14:42:07Z" level=info msg="Collecting Pod Logs by FieldSelectors []"
time="2021-11-04T14:42:07Z" level=info msg="Log lines after this point will not appear in the downloaded tarball."
time="2021-11-04T14:42:07Z" level=info msg="Results available at /tmp/sonobuoy/202111041439_sonobuoy_dae67975-cdc1-488c-b11f-81ae10346043.tar.gz"
time="2021-11-04T14:42:07Z" level=info msg="no-exit was specified, sonobuoy is now blocking"
```

Looks like there is some encoding/decoding problem with the e2e results:

```
time="2021-11-04T14:39:55Z" level=error msg="Result processing error (500): error handling result e2e/global: couldn't decode result e2e/results/global: couldn't uncompress reader: gzip: invalid header"
time="2021-11-04T14:39:56Z" level=error msg="Result processing error (500): error handling result e2e/global: couldn't decode result e2e/results/global: couldn't uncompress reader: gzip: invalid header"
time="2021-11-04T14:39:57Z" level=error msg="Result processing error (500): error handling result e2e/global: couldn't decode result e2e/results/global: couldn't uncompress reader: gzip: invalid header"
time="2021-11-04T14:40:04Z" level=info msg="All plugins have completed, status has been updated"
```

Environment:

Sonobuoy version: 0.54.0
Kubernetes version: (use kubectl version): 1.19.15
Tarball: [results.tar.gz](https://github.com/vmware-tanzu/sonobuoy/files/7481366/results.tar.gz)

If you need any additional logs please let me know.

**Comment by johnSchnake** [comment]
```
time="2021-11-04T14:39:57Z" level=error msg="Result processing error (500): error handling result e2e/global: couldn't decode result e2e/results/global: couldn't uncompress reader: gzip: invalid header"
```

I just saw this myself yesterday but I can't recall if I was on v0.54 or a dev build.

I definitely want to look into this and figure out/resolve the bug before a next release.

**Comment by johnSchnake** [comment]
```
time="2021-11-04T14:39:57Z" level=error msg="Result processing error (500): error handling result e2e/global: couldn't decode result e2e/results/global: couldn't uncompress reader: gzip: invalid header"
```

I just saw this myself yesterday but I can't recall if I was on v0.54 or a dev build.

I definitely want to look into this and figure out/resolve the bug before a next release.

**Comment by johnSchnake** [comment]
If you have a full tarball I'd love to look at it.

Locally I'm not able to repro right now; trying with the same sonobuoy and conformance versions and not having luck reproducing right now.

**Comment by mauricewittek** [comment]
I just added the tarball. I had to remove some files and hostnames/ip-addresses. 

Hopefully the files are still usefull for you.

**Comment by johnSchnake** [comment]
So its very confusing to keep all these bits straight but I can see the gist of the issue.

If you look at the e2e plugin pod there are two containers as expected, the e2e container and the sonobuoy-worker.

The sonobuoy-worker looks to be in good shape in general. It has a SONOBUOY_RESULTS_DIR and RESULTS_DIR and volume mount that all agree on the results going into `/tmp/sonobuoy` which is what your config file says.

The e2e container is a different story:
 - Volume mount is `/tmp/results`
 - SONOBUOY_RESULTS_DIR is `/tmp/sonobuoy/results` (default value, not from config)
 - no RESULTS_DIR at all
 - https://github.com/kubernetes/kubernetes/blob/d5352acc4821e8ffec1356a3e8f8d908d680ee88/test/conformance/image/go-runner/const.go#L29 shows that the upstream image is looking specifically for RESULTS_DIR

So the flow is a bit hard to follow but I think the issue is just that the plugin container is ignoring the SONOBUOY_RESULTS_DIR, looking for RESULTS_DIR, not finding it, and using what the upstream considers the default, `/tmp/results` https://github.com/kubernetes/kubernetes/blob/dbebb8f2e8361db92ccc72ab9da44d7e5b38f76f/test/conformance/image/go-runner/const.go#L60

The sonobuoy-worker is waiting for the 'donefile' which it thinks is at `waitfile=/tmp/sonobuoy/done`; it reports seeing it but I have no idea why. The plugin never mentioned wanting to write to `/tmp/sonobuoy`. However, the worker then gets the correct name of the results file `resultFile=/tmp/results/e2e.tar.gz` and tries to submit it. However, it doesnt have the `/tmp/results` mount causing it to not find the tarball and sending empty/junk data to the aggregator, getting an error, and ultimately not knowing the results.

Solution:
 - PRs have already merged to fix this sort of thing; it ensures that the plugins/workers agree about the results mount and it comes from the config as expected
 - This fix isn't available until v0.55.0 unless you want to manually build sonobuoy (you're already using the `main` image but not the 'main'/tip of the CLI.

Workaround without waiting for the release:
 - Do a `sonobuoy gen ...` of your run command to see the yaml its going to submit to the server. Modify the e2e plugin to have the `RESULTS_DIR=/tmp/sonobuoy` and the same volume mount. This should make it agree with the sonobuoy-worker and things should go smoothly.

TBH I don't know how we didnt have more of these types of issues before since I found the config.ResultsDir wasn't being appropriately passed around and different pieces of code seemed to make different assumptions about hardcoding path values. Sorry you're having to deal with those oversights.

**Comment by mauricewittek** [comment]
I just tested the workaround and it works fine, now I can retrive the results of the e2e test as well:

```
sonobuoy results 202111051621_sonobuoy_150a34d6-f8d2-48fc-926b-a41a02822ebf.tar.gz
Plugin: e2e
Status: passed
Total: 5484
Passed: 1
Failed: 0
Skipped: 5483

Plugin: systemd-logs
Status: passed
Total: 3
Passed: 3
Failed: 0
Skipped: 0
```

Thanks for your support, have a nice day. 🥇

**Comment by yehaha9876** [comment]
I have same issue，here is  custom-repos.yaml example
```
gcAuthenticatedRegistry: "registry.gzky.com:8443/authenticated-image-pulling"
promoterE2eRegistry: "registry.gzky.com:8443/e2e-test-images"
buildImageRegistry: "registry.gzky.com:8443/build-image"
invalidRegistry: "registry.gzky.com:8443/invalid"
gcEtcdRegistry: "registry.gzky.com:8443"
gcRegistry: "registry.gzky.com:8443"
sigStorageRegistry: "registry.gzky.com:8443/sig-storage"
privateRegistry: "registry.gzky.com:8443/k8s-authenticated-test"
microsoftRegistry: "registry.gzky.com:8443"
dockerLibraryRegistry: "registry.gzky.com:8443/library"
cloudProviderGcpRegistry: "registry.gzky.com:8443/cloud-provider-gcp"
```

it came from env , read this:
https://github.com/kubernetes/kubernetes/tree/release-1.22/test/e2e
